torch/csrc/jit/api/module.cpp

```diff
@@ -118,6 +118,17 @@ IValue Method::operator()(std::vector<IValue> stack, const Kwargs& kwargs) {
   return (*function_)(std::move(stack), kwargs);
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> Method::run_async(
+    std::vector<IValue> stack,
+    const Kwargs& kwargs,
+    TaskLauncher taskLauncher) {
+  stack.insert(stack.begin(), owner()._ivalue());
+  RECORD_TORCHSCRIPT_FUNCTION(name(), stack);
```

Comment:
@iliacher to see whether the logging keeps working as expected (because the destructor of RecordFunction will be called before the function really finishes)

Reply:
yep it won't work this way - you'd only log the async launch, not the method execution itself

Reply:
any suggestions on how to fix it? I guess we need to construct RecordFunction as unique_ptr and destroy it only when the future is satisfied. Will it work if it happens on a different thread?

alternatively if we keep it as-is, would it be a problem for profiling? I guess it might cause the event ranges not perfectly nested

I guess in general we should have a coherent story how profiler ranges should work with multithreading

Reply:
yes, RecordFunction is ok with finishing on another thread, but can we instead just trace the actual async job? if you absolutely have to trace the whole [launch, end of async job] span then you could move Recordfunction into the async lambda (e.g. async job, or a callback attached to async job future)

Reply:
(by move I mean, create e.g. a unique_ptr in launch section and move it into the lambda)

Reply:
I'd really really encourage not to do this though

@dzhulgakov  re. consistency - the only use case that does this is in distributed profiling code, and analyzing these async (start on one thread, end on another) ranges is a problematic, e.g. profiler wouldn't report much about them

so I'd say, unless this is absolutely needed, we don't use RecordFunction across thread boundaries

Reply:
another use case is when code in TS interpreter calls wait, and then wakes up in another thread, in this case we get an async range too, but that's mostly unavoidable, unless you want to finish recording when code calls wait

Reply:
So, should we keep as it is? Today's "forked" function will be executed in another thread, has its own record function, which is unrelated to the original one (return immediately with future).

Reply:
Yeah, I guess we can keep it as is for now then. It won't attribute continuations to the same original invocation, but it seems that profiler is not too good at visualizing those anyway.